### PR TITLE
Implement extract_features

### PR DESCRIPTION
**Pull request by m09**
The parent detection is at the moment very strict. It will be important to test different methods (such as lowest common ancestor or some other method).

Signed-off-by: Hugo Mougard <hugo@sourced.tech>

### CODE REVIEW

File: lookout/style/format/features.py [hunk 1]
```diff
@@ -99,17 +98,40 @@ class FeatureExtractor:
...
-        :return:
+        :param parents_depth: how many parents to use for each node.
+        :param siblings_window: how many siblings to use for each node (both left and right).
+        :return: tuple of numpy.ndarray (2 and 1 dimensional respectively): features and labels.
         """
-        pass
+        parsed_files = []
+        labels = []
+        for file in files:
+            contents = file.content.decode('utf-8')
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: We use double quotes https://github.com/src-d/guide/blob/master/engineering/conventions-python.md#code-style

Comment (vmarkovtsev) [reply lookout/style/format/features.py]: From my experience we will definitely hit UnicodeDecodeError-s here.

```python
content.decode("utf-8", "replace")
```

The replacement is needed to not interfere with the offsets.

File: lookout/style/format/features.py [hunk 2]
```diff
@@ -99,17 +98,40 @@ class FeatureExtractor:
     def __init__(self, language: str):
         self.tokens = importlib.import_module("lookout.style.format.langs.%s.tokens" % language)
         self.roles = importlib.import_module("lookout.style.format.langs.%s.roles" % language)
+        self.feature_names = ['start_offset', 'start_line', 'start_col', 'end_offset', 'end_line',
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: This array seems to be constant so may be reasonable to move to the class level

File: lookout/style/format/features.py [hunk 3]
```diff
@@ -99,17 +98,40 @@ class FeatureExtractor:
...
+        for file in files:
+            contents = file.content.decode('utf-8')
+            uast = file.uast
+            vnodes, parents = self._parse_file(contents, uast)
+            vnodes = self._classify_vnodes(vnodes)
+            parsed_files.append((vnodes, parents))
+            labels.append([vnode.y for vnode in vnodes if vnode.y])
+
+        y = numpy.concatenate(labels)
+
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: pls remove this empty line

File: lookout/style/format/features.py [hunk 4]
```diff
@@ -225,17 +247,95 @@ def _classify_vnodes(self, nodes: Iterable[VirtualNode]) -> List[VirtualNode]:
                             y=cls))
         return result
 
+    def _inplace_write_vnode_features(self, features: numpy.ndarray, vnode: VirtualNode, row: int,
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: The parameters order should be first input then output (e.g. `_inplace_write_vnodes_features`)

File: lookout/style/format/features.py [hunk 5]
```diff
@@ -225,17 +247,95 @@ def _classify_vnodes(self, nodes: Iterable[VirtualNode]) -> List[VirtualNode]:
...
+                                      vnode_index: int) -> None:
+        """
+        Given a feature `numpy.ndarray`, a `VirtualNode`, a row and a feature index, write the
+        features of the vnode at the correct location in the feature matrix.
+
+        :param features: `numpy.ndarray` the features matrix.
+        :param vnode: `VirtualNode` we want to write the features of.
+        :param row: on which row to write the features.
+        :param vnode_index: the column at which we'll start writing the features is
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: Sphinx requires `\` when breaking the line for `:param` (if I remember correctly)

File: lookout/style/format/features.py [hunk 6]
```diff
@@ -225,17 +247,95 @@ def _classify_vnodes(self, nodes: Iterable[VirtualNode]) -> List[VirtualNode]:
...
+        :param parents: dictionnary of node id to parent node
+        :param parents_depth: number of parents to use in the features
+        :param siblings_window: number of siblings to use, both at the left and right of the node
+        :param line_offset: at which line of the input ndarrays should we start writing
+        :param X: features matrix
+        """
+        last_left_parent = None
+        i = 0
+        for vnode in vnodes:
+
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: pls remove this empty new line

File: lookout/style/format/features.py [hunk 7]
```diff
@@ -225,17 +247,95 @@ def _classify_vnodes(self, nodes: Iterable[VirtualNode]) -> List[VirtualNode]:
...
+            # compute current vnode parent. If both left and right closest parent are the same then
+            # we use it else we use no parent feature (set them to -1 later on)
+            if vnode.node:
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: Makes sense to extract this into a function

File: lookout/style/format/features.py [hunk 8]
```diff
@@ -225,17 +247,95 @@ def _classify_vnodes(self, nodes: Iterable[VirtualNode]) -> List[VirtualNode]:
...
+
+            # compute how many dummy features we'll need to account for the lack of left and right
+            # siblings and the lack of parents
+            n_dummies_left = abs(min(i - siblings_window, 0))
+            parents_list = []
+            if parent:
+                current_vnode = vnode
+                for j in range(parents_depth):
+                    node_id = id(current_vnode)
+                    if node_id in parents:
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: consider

```python
if node_id not in parents:
    break
parent = parents[node_id]
...
```

File: lookout/style/format/features.py [hunk 9]
```diff
@@ -225,17 +247,95 @@ def _classify_vnodes(self, nodes: Iterable[VirtualNode]) -> List[VirtualNode]:
...
+                        parent = parents[node_id]
+                        parents_list.append(parent)
+                        current_vnode = parent
+                    else:
+                        break
+            self._inplace_write_vnode_features(X, vnode, i + line_offset, 0)
+            for j, left_vnode in enumerate(vnodes[max(0, i - siblings_window):i]):
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: could you please put a comment here...

what do we write for each node. Window left, window right, parent... `parents_list`?

File: lookout/style/format/features.py [hunk 10]
```diff
@@ -1,12 +1,11 @@
 import importlib
-from typing import List, Dict, Tuple, NamedTuple, Iterable
+from typing import Dict, Iterable, List, Mapping, NamedTuple, Sequence, Tuple
 
-import bblfsh
 import numpy
 
+import bblfsh
```
[the remark this answers is not in the document]
Comment (m09) [reply lookout/style/format/features.py]: weird, I installed some atom extension using isort (that has as only goal to sort imports according to PEP8) and it did this indeed wrong sort. I'll revert to checking them manually :(

[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply lookout/style/format/features.py]: PyCharm has Code -> Optimize Imports which can be triggered on each save

File: lookout/style/format/tests/test_features.py [hunk 11]
```diff
@@ -1,12 +1,14 @@
 import lzma
+import unittest
 from collections import defaultdict
 from pathlib import Path
-import unittest
```
Comment (vmarkovtsev) [inline lookout/style/format/tests/test_features.py]: same problem with the imports here haha

File: lookout/style/format/features.py [hunk 12]
```diff
@@ -225,17 +254,114 @@ def _classify_vnodes(self, nodes: Iterable[VirtualNode]) -> List[VirtualNode]:
...
+        if vnode.node:
+            role = vnode.node.internal_type
+            if role in self.roles.ROLE_INDEX:
+                role_index = self.roles.ROLE_INDEX[role]
+        col = vnode_index * len(self.feature_names)
+        features[row, col:col + 4] = pos
+        features[row, col + 4] = role_index
+
+    @staticmethod
+    def _assign_parent(vnode: VirtualNode, vnode_index: int, vnodes: Sequence[VirtualNode],
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: better to rename this to `_find_parent()` - we do not really assign anything inside.

File: lookout/style/format/features.py [hunk 13]
```diff
@@ -225,17 +254,114 @@ def _classify_vnodes(self, nodes: Iterable[VirtualNode]) -> List[VirtualNode]:
...
+                       parents: Mapping[int, bblfsh.Node], last_left_parent: bblfsh.Node
+                       ) -> Optional[bblfsh.Node]:
+        # compute current vnode parent. If both left and right closest parent are the same then
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: docstring instead of comments

**Pull request merged**